mapping_pcd:
  # Topics
  # Xe 4x4
  # pointCloudTopic: "/sensing/lidar/top/pointcloud_raw_ex"                # Point cloud data
  # imuTopic: "/sensing/imu/tamagawa/imu/data"   #"/sensing/imu/tamagawa/imu/data"   
  # gpsTopic: "/sensing/gnss/ublox/nav_sat_fix"                         # GPS data
  
  # Xe Vinfast
  # pointCloudTopic: "/points_raw"                                    # Point cloud data
  # imuTopic: "/imu/data"                                             # IMU data
  # gpsTopic: "/piksi/navsatfix_best_fix"                             # GPS data

  # Balo map
  pointCloudTopic: "/points_raw"                                    # Point cloud data
  imuTopic: "/imu/data"                                             # IMU data
  gpsTopic: "/fix"                                                  # GPS data
  
  odomTopic: "odometry/imu"                   # IMU pre-preintegration odometry, same frequency as IMU
  gpsOdom_Topic: "odometry/gps"                   # GPS odometry topic from navsat

  # Frames
  # lidarFrame: "velodyne_top"
  # Balo map
  lidarFrame: "velodyne"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"
  map_project: mgrs

  # GPS Settings
  useImuHeadingInitialization: false       
  useGpsElevation: true                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 1.0                       # m^2, threshold for using GPS data 
  gpsCovThreshold_z: 1.5                        # m, threshold for z axis GPS
  poseCovThreshold: 0.1                      # m^2, threshold for using GPS data
  useNavsat: false
  
  # Export settings
  savePCD: true                           # https://github.com/TixiaoShan/LIO-SAM/issues/3
  savePCDDirectory: "/Downloads/LOAM/"        # in your home folder, starts and ends with "/". Warning: the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation
  loadmapDirectory: "/Data_mapping_Velodyne_VLP32C/Ecopark/data_07_01_2022/01/"                      #"/Autoware_IV_file/map_folder/"

  # Sensor Settings
  sensor: velodyne                       # lidar sensor type, either 'velodyne' or 'ouster' or 'pandar'
  N_SCAN: 32                               # number of lidar channel (i.e., 16, 32, 64, 128)
  Horizon_SCAN: 1800                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048)
  downsampleRate: 1.0                         # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
  lidarMinRange: 1.0                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 1000.0                       # default: 1000.0, maximum lidar range to be used
  useCloudRing: true
  ang_res_x: 0.2                              #Goc cua cung thuoc 1 channel: 360.0/Horizon_SCAN        Velodyne 16, 32, 64: 0.2  
  ang_res_y: 1.29032                          #                              40/(N_SCAN-1)             Velodyne 16: 2.0 - Velodyne VLP32C: 1.29032
  ang_bottom: 25.0                            #15.0
  groundScanInd: 20                           #8

  # IMU Settings (Balo map) __ Bản 1 __ XSENS_MTI-10      # Ngày 27/11/2021 - VLP32C
  imuAccNoise: 3.2399643052664455e-02   #acc_n
  imuGyrNoise: 1.0870558780580436e-02    #gyr_n
  imuAccBiasN: 3.5516567551560147e-04   #acc_w
  imuGyrBiasN: 7.9520395862518761e-05    #gyr_w

  # IMU Settings (Balo map) __ Bản 2 __ XSENS MTI-19      # Ngày 08/10/2021 - VLP16
  # imuAccNoise: 1.9224062852804712e-02   #acc_n
  # imuGyrNoise: 1.0839388774796179e-02    #gyr_n
  # imuAccBiasN: 3.8758397373734997e-04   #acc_w
  # imuGyrBiasN: 9.8751635519424619e-05    #gyr_w

  # IMU Settings (Default)
  # imuAccNoise: 3.9939570888238808e-03
  # imuGyrNoise: 1.5636343949698187e-03
  # imuAccBiasN: 6.4356659353532566e-05
  # imuGyrBiasN: 3.5640318696367613e-05

  # IMU Settings (Xe dien)
  # imuAccNoise: 6.2631420422441975e-01    #acc_n
  # imuGyrNoise: 2.3943694918095293e-01    #gyr_n
  # imuAccBiasN: 1.5823925722404289e-01    #acc_w
  # imuGyrBiasN: 2.6982260728014539e-03   #gyr_w

  imuRate: 400.0  
  imuGravity: 9.80511 
  imuRPYWeight: 0.01

  # Extrinsics (lidar -> IMU)
# Xe dien 2x2
  # extrinsicTrans: [1.23791, 0.0, 1.47866]

# Xe VINFAST
  # extrinsicTrans: [0.5, 0.0, 1.0]
  # extrinsicRot: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]
  # extrinsicRPY: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]

# Balo map
  extrinsicTrans: [0.0, 0.0, 0.043]  #Lidar theo base link (0, 0, 0.497, 0, 0, 0)
  extrinsicRot: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]
  extrinsicRPY: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]

  #Xe dien tung lam:
  # extrinsicRot: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]
  # extrinsicRPY: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]  

# Xe 4x4
  # extrinsicTrans: [0.479, 0.0, 0.063]   #Lidar theo base link (0.953 0 2.043 0 0 0)
  # extrinsicRot: [1, 0, 0,
  #                0, 1, 0,
  #                0, 0, 1]
  # extrinsicRPY: [1, 0, 0,
  #                 0, 1, 0,
  #                 0, 0, 1]

  # LOAM feature threshold
  edgeThreshold: 1.0
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10     #80 - indoor, 100 - outdoor
  surfFeatureMinValidNum: 100     #160 - indoor, 500 - outdoor

  # Filter param (Đơn vị meters)
  odometrySurfLeafSize: 0.4                  # default: 0.4 - outdoor, 0.2 - indoor, 0.06 - indoor_balo, 0.1 - outdoor_balo
  mappingCornerLeafSize: 0.2                   # default: 0.2 - outdoor, 0.1 - indoor, 0.03 - indoor_balo, 0.05 - outdoor_balo
  mappingSurfLeafSize: 0.4                   # default: 0.4 - outdoor, 0.2 - indoor, 0.06 - indoor_balo, 0.1 - outdoor_balo
  kSCLeafSize: 0.5                            # 0.03 - indoor_balo, 0.05 - outdoor_balo, 0.5 - default
  kDownsampleVoxelSize: 0.05                 # Hệ số cho Octree filter

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 16                             # number of cores for mapping optimization
  mappingProcessInterval: 0.15                 # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0  # meters, regulate keyframe adding threshold, 1.0 - 0.1
  surroundingkeyframeAddingAngleThreshold: 0.2 # radians, regulate keyframe adding threshold, 0.2 - 0.02
  surroundingKeyframeDensity: 2.0              # meters, downsample surrounding keyframe poses, 0.05 - outdoor  
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Scancontext loop
  loopClosureSCEnableFlag: true
  mode_SC: normal                  # normal or intensity 
  pc_dist_thresh: 0.4 
  pc_max_radius: 80.0
  historyKeyframeFitnessScore_SC: 0.3 

  # Loop closure
  loop_algoth: icp                   # Chon thuat toan matching/loop 'icp' hoac 'gicp' hoac 'ndt'
  loopClosureEnableFlag: true
  loopClosureFrequency: 2.0                  # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 60                   # submap size (when loop closure enabled) 30*2=60
  historyKeyframeSearchRadius: 80.0  #50.0    #15.0 - indoor, 50.0 - outdoor            # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 15.0   #8.0   #30.0 - indoor, 30.0 - outdoor               # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 30                    # number of history key frames will be fused into a submap for loop closure [-30, 30]
  historyKeyframeFitnessScore_RS: 1.2             # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 10000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 0.2      # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 0.2           # meters, global map visualization cloud density

# Navsat (convert GPS coordinates to Cartesian)
navsat:
  frequency: 50
  wait_for_datum: false #true
  delay: 0.0
  magnetic_declination_radians: 0
  yaw_offset: 0
  zero_altitude: false
  use_odometry_yaw: false #true
  broadcast_utm_transform: false
  broadcast_utm_transform_as_parent_frame: false
  publish_filtered_gps: true #false 

# EKF tính toán cho IMU và GPS tạo ra Odom cho mapping (Không kết hợp thêm Odom nào ngoại trừ gps)
ekf_gps:
  publish_tf: false
  map_frame: map
  odom_frame: odom
  base_link_frame: base_link
  world_frame: odom

  frequency: 50
  two_d_mode: false
  sensor_timeout: 0.01

  # -------------------------------------
  # External IMU:
  # -------------------------------------
  # Xe 4x4
  # imu0: /sensing/imu/tamagawa/imu/data    #/sensing/imu/tamagawa/imu/data
  
  # Xe Vinfast và balo map
  imu0: /imu/data
 
  # input is aligned with ROS REP105. EKF can also transform the data using tf between your imu and base_link
  imu0_config: [false, false, false,   #x, y, z
                true,  true,  true,   #roll, pitch, yaw
                false, false, false,  #x velocity, y velocity, z velocity
                true, true, true,   #roll velocity, pitch velocity, yaw velocity (Goc cua van toc)
                true,  true,  true]   #x accel, y accel, z accel
  imu0_differential: false
  imu0_queue_size: 50 
  imu0_remove_gravitational_acceleration: true

  # -------------------------------------
  # Odometry (From Navsat) -- GPS:
  # -------------------------------------
  odom0: /odometry/gps               #odometry xuat ra tu Navsat, dau vao EKF
  odom0_config: [true,  true,  true,   #x, y, z
                 false, false, false,  #roll, pitch, yaw
                 false, false, false,   #x velocity, y velocity, z velocity
                 false, false, false,  #roll velocity, pitch velocity, yaw velocity (Goc cua van toc)
                 false, false, false]  #x accel, y accel, z accel (Gia toc)
  odom0_differential: false
  odom0_queue_size: 10

                               x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
  process_noise_covariance: [  1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015]

  # initial_estimate_covariance: [1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
  #                               0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
  #                               0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
  #                               0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
  #                               0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
  #                               0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
  #                               0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
  #                               0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
  #                               0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
  #                               0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
  #                               0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
  #                               0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
  #                               0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0,
  #                               0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0,
  #                               0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]
